src/utils/config_filters.hpp

```diff
@@ -40,8 +40,8 @@ bool bool_matches_if_present(const config& filter, const config& cfg, const std:
  *
  * Always returns true if the filter puts no restriction on the value of @a cfg[@a attribute].
  */
-bool double_matches_if_present(const config& filter, const config& cfg, const std::string& attribute, std::optional<double> def = NULL);
-bool int_matches_if_present(const config& filter, const config& cfg, const std::string& attribute, std::optional<int> def = NULL);
+bool double_matches_if_present(const config& filter, const config& cfg, const std::string& attribute, std::optional<double> def = std::nullopt);
```

Comment:
Will this constitute a behavior change?
In before, when `def` is absence, an `std::optional` which will constructed from `NULL` (in reality, it's `0`?). 
Then https://github.com/wesnoth/wesnoth/blob/c1b331c62fd1c4d14d83da5bb88592bb7ce518ce/src/utils/config_filters.cpp#L98
will see `def` converted to `true` (value presents), thus `value_def` will be `0`.

With this change, `value_def` is `1` when `def` is absence.

Were I wrong?

Reply:
When def is absent and cfg attr both absent return false before value_def defined

Reply:
No, but ithink what it was already like this before to aknowledge my error

Reply:
Well, I have no idea :shrug:

Reply:
I don't think `NULL` is 0? Usually it's defined as `(void*)0` is it not? So it has type `void*`, which shouldn't be implicitly convertible to a number.

In any case, even if there's a behaviour change here, the old behaviour must be a bug.

Reply:
Exact

> I don't think `NULL` is 0? Usually it's defined as `(void*)0` is it not? So it has type `void*`, which shouldn't be implicitly convertible to a number.
> 
> In any case, even if there's a behaviour change here, the old behaviour must be a bug.

Exactly

Reply:
> I don't think `NULL` is 0? Usually it's defined as `(void*)0` is it not? So it has type `void*`, which shouldn't be implicitly convertible to a number.

[Maybe, maybe not](https://devblogs.microsoft.com/oldnewthing/20180307-00/?p=98175), maybe it depends on whether `#include <cstdlib>` comes before or after any [SDL headers](https://github.com/libsdl-org/SDL/blob/0d8ce4a761a8e84e5b746d0d53e7b1dd8fc92d4d/include/begin_code.h#L159-L168).

PR #8651 is trying to get an answer for the platforms that we build on, but we don't build those tests on mingw.

Reply:
Even if this is a behavior change, one should always use `nullopt` for a default optional value, not any value that could potentially lead to it "having" an unintended variable (like 0) (especially something like `optional<int>`). Given the old code was *assuming* that it would default to "no value", if it doesn't handle that case properly it would be a bug in the implementation (that is, if the "has value" codepath was executed with the default value we'd have a problem).

Reply:
OK i will remove default_value

Reply:
done

Reply:
> I don't think `NULL` is 0?

No, only `0` is a correct value for `NULL` if we use the same definition for both C and C++, since C++ doesn't allow implicit cast from `void *` to other pointer types. But there's other caveats, thus it's complicated. 

> Usually it's defined as `(void*)0` is it not?

No, [musl defines it](https://git.musl-libc.org/cgit/musl/tree/include/stddef.h#n4) to:
```cpp
#if __cplusplus >= 201103L
#define NULL nullptr
#elif defined(__cplusplus)
#define NULL 0L
#else
#define NULL ((void*)0)
#endif
```

glibc outsourced that to gcc, and [yes at some points](https://github.com/gcc-mirror/gcc/blob/26a723625bb63e42b8c78bbce52ba0286eefda1c/gcc/ginclude/stddef.h#L401) it's `0`:
```cpp
#if defined (_STDDEF_H) || defined (__need_NULL)
#undef NULL		/* in case <stdio.h> has defined it. */
#ifdef __GNUG__
#define NULL __null
#else   /* G++ */
#ifndef __cplusplus
#define NULL ((void *)0)
#else   /* C++ */
#define NULL 0
#endif  /* C++ */
#endif  /* G++ */
#endif	/* NULL not defined and <stddef.h> or need NULL.  */
#undef	__need_NULL
```

Reply:
Anyway, I just want to point out behaviour changes, I'm not against any changes, though.